accel-config: Clean up memory leak, errouneous assignment, and unused path variable

snits opened the pull request:
The following pattern occurs with wq_base, group_base, and engine_base in accfg/lib/libaccfg.c:

    // Where X is one of: wq, group, engine
    X_base_string = strdup(X_base);
    sscanf(X_base_string,...);
    ...
    X->X_path = strdup(X_base);

    and X_base_string is never cleaned up with free().

Instead of calling strdup() twice, just assign X_base_string to X->X_path, clean up the error checking related to the 2nd strdup(), and make the error path clean up more consistent between the cases.

Also a really minor clean up of test in accfg/test.c, so covscan stops complaining about it.

Signed-off-by: Jerry Snitselaar <jsnitsel@redhat.com>

snits committed:
accel-config: Clean up unused path variable.

add_wq, add_group, and add_engine all have a path
variable that they calloc, and then never use,
freeing it at the end of the function before returning.

Signed-off-by: Jerry Snitselaar <jsnitsel@redhat.com>

snits commented:
@ramesh-thomas I added that 2nd commit later when I noticed that it doesn't appear that anything is making use of that path variable. So not tied to the memory leak complaints from coverity, but I figured I would just tack it on this merge request.

snits commented:
Darn it. I need to fix something up that I missed in the conflict resolution.

ramesh-thomas commented:
> Darn it. I need to fix something up that I missed in the conflict resolution.

Sure. Let me know when you are done. I will merge them and make a new release. Thanks.

snits commented:
> > Darn it. I need to fix something up that I missed in the conflict resolution.
> 
> Sure. Let me know when you are done. I will merge them and make a new release. Thanks.

I think it should be good now.

ramesh-thomas commented on accfg/lib/libaccfg.c:
```diff
@@ -835,11 +833,7 @@ static void *add_wq(void *parent, int id, const char *wq_base,
-		err(ctx, "forming of wq path failed\n");
-		goto err_read;
-	}
+	wq->wq_path = wq_base_string;
```
Reason a copy of wq_base was used is because basename() can modify the input path parameter. 
https://man7.org/linux/man-pages/man3/basename.3.html

If we can assume that basename() wouldn't modify the input path then we don't need wq_base_string at all.

Otherwise we can do strcpy(wq_base_string, wq_base) to put back the original path in wq_base.  Maybe add a comment so it would be clear, like "/* basename() could have modified the path */"

snits replied:
> Reason a copy of wq_base was used is because basename() can modify the input path parameter. https://man7.org/linux/man-pages/man3/basename.3.html
> 
> If we can assume that basename() wouldn't modify the input path then we don't need wq_base_string at all.
> 
> Otherwise we can do strcpy(wq_base_string, wq_base) to put back the original path in wq_base. Maybe add a comment so it would be clear, like "/* basename() could have modified the path */"

Ugh, yep. Let me fix that up adding those bits back in, and I'll add a comment to make the basename part clear.

ramesh-thomas commented on accfg/lib/libaccfg.c:
```diff
@@ -900,17 +895,15 @@ static void *add_group(void *parent, int id, const char *group_base,
 	}
-	free(group_base_string);
-
-	group->group_path = (char *)group_base;
```
This was a redundant and erroneous assignment. Thanks for removing.

ramesh-thomas commented on accfg/lib/libaccfg.c:
```diff
@@ -900,17 +895,15 @@ static void *add_group(void *parent, int id, const char *group_base,
-	free(group_base_string);
-
-	group->group_path = (char *)group_base;
+	group->group_path = group_base_string;
```
strcpy(group_base_string, group_base); /* basename() could have modified path */

ramesh-thomas commented on accfg/lib/libaccfg.c:
```diff
@@ -994,30 +983,22 @@ static void *add_engine(void *parent, int id, const char *engine_base,
-		err(ctx, "forming of engine path failed\n");
-		goto err_read;
-	}
+	engine->engine_path = engine_base_string;
```
strcpy(engine_base_string, engine_base); /* basename() could have modified path */

snits committed:
accel-config: Clean up memory leak, and erroneous assignment

Clean up memory leak in cmd_test(), and erroneous assignment
to group->group_path in add_group().

Signed-off-by: Jerry Snitselaar <jsnitsel@redhat.com>

snits commented:
I dropped the error path changes for now, and just pushed the fix up of the leak and erroneous assignment, plus the path variable clean up. I'll look at doing a re-organization of those error paths separately if you would like that.

snits committed:
accel-config: Make error paths more consistent for add_wq, add_group, add_engine

Signed-off-by: Jerry Snitselaar <jsnitsel@redhat.com>

snits commented:
> I dropped the error path changes for now, and just pushed the fix up of the leak and erroneous assignment, plus the path variable clean up. I'll look at doing a re-organization of those error paths separately if you would like that.

And then after dinner, and walking the dog I decided to stick it on top as a separate patch. :) @ramesh-thomas let me know if you want me to pull that from this pr.

ramesh-thomas commented:
> And then after dinner, and walking the dog I decided to stick it on top as a separate patch. :) @ramesh-thomas let me know if you want me to pull that from this pr.

I think your patch improves the readability. The previous code, though not broken, was relying too much on the fact that free(NULL) is a nop and was doing redundant freeing. I have merged it into pending. Will move to stable once the location of the config files are finalized.
